contracts/non-fungible-tokens/andromeda-crowdfund/src/contract.rs

```diff
@@ -345,63 +346,66 @@ fn handle_receive_cw20(
     }
 }
 
-// fn execute_end_campaign(ctx: ExecuteContext) -> Result<Response, ContractError> {
-//     let ExecuteContext {
-//         deps, info, env, ..
-//     } = ctx;
+fn execute_end_campaign(ctx: ExecuteContext) -> Result<Response, ContractError> {
+    nonpayable(&ctx.info)?;
 
-//     // Only owner can end the campaign
-//     let contract = ADOContract::default();
-//     ensure!(
-//         contract.is_contract_owner(deps.storage, info.sender.as_str())?,
-//         ContractError::Unauthorized {}
-//     );
+    let ExecuteContext {
+        deps, info, env, ..
+    } = ctx;
 
-//     // Campaign is finished already or not started
-//     let curr_stage = get_current_stage(deps.storage);
-//     ensure!(
-//         curr_stage == CampaignStage::ONGOING,
-//         ContractError::InvalidCampaignOperation {
-//             operation: "end_campaign".to_string(),
-//             stage: curr_stage.to_string()
-//         }
-//     );
+    // Only owner can end the campaign
+    let contract = ADOContract::default();
+    ensure!(
+        contract.is_contract_owner(deps.storage, info.sender.as_str())?,
+        ContractError::Unauthorized {}
+    );
 
-//     let current_cap = get_current_cap(deps.storage);
-//     let campaign_config = get_config(deps.storage)?;
-//     let soft_cap = campaign_config.soft_cap.unwrap_or(Uint128::one());
-//     let end_time = campaign_config.end_time;
-
-//     // Decide the next stage
-//     let next_stage = match (current_cap >= soft_cap, end_time.is_expired(&env.block)) {
-//         (true, _) => CampaignStage::SUCCESS,
-//         (false, true) => CampaignStage::FAILED,
-//         (false, false) => {
-//             if current_cap != Uint128::zero() {
-//                 return Err(ContractError::CampaignNotExpired {});
-//             }
-//             CampaignStage::ONGOING
-//         }
-//     };
+    // Campaign is finished already or not started
+    let curr_stage = get_current_stage(deps.storage);
+    ensure!(
+        curr_stage == CampaignStage::ONGOING,
```

Comment:
I think the owner should be able to discard the sale on success as well, in the case there is a problem with the sale recipient. In this case the tokens may become worthless and are still claimable and users should be able to receive a refund as if the sale failed.

Reply:
This was discussed in our planning call and this is what we decided upon if I remember correctly.

CC: @SlayerAnsh

Reply:
I am not sure if that kind of logic is necessary.
Currently, `withdrawal_recipient` is set only in the `instantiate` function and `withdrawal_address` is validated.
```
pub fn instantiate(
    deps: DepsMut,
    env: Env,
    info: MessageInfo,
    msg: InstantiateMsg,
) -> Result<Response, ContractError> {
    let branch = DepsMut {
        storage: deps.storage,
        api: deps.api,
        querier: deps.querier,
    };
    msg.campaign_config.validate(branch, &env)?;
    update_config(deps.storage, msg.campaign_config)?;
...  
```
```
impl CampaignConfig {
    pub fn validate(&self, deps: DepsMut, env: &Env) -> Result<(), ContractError> {
        // validate addresses
        self.tier_address.validate(deps.api)?;
        self.withdrawal_recipient.validate(&deps.as_ref())?;
...
```

Thus the campaign is always instantiated with valid recipient and there will be no problem with the withdrawal recipient to end the campaign.

Reply:
The recipient can include an extra `ExecuteMsg` for another contract but we have no way of validating that the message is correct. If this errors then the owner of the contract cannot withdraw the proceeds as expected.

Reply:
Added additional `is_discard` field to the `EndCampaign` message for discard functionality.

Reply:
CC: @SlayerAnsh